File: api/sos/engine.py
"""The situation engine: one pure function from the model to the View that every screen reads.

`compute(model, rules)` takes the conditions, the scenario clock, the home and the time, and returns the View:
effective conditions, inferred proposals, a forecast with due times, the task list, what to read, the interface
modes and the next radio bulletin. No I/O, no clock of its own, no randomness: the same model and the same rules
always give the same View, which is what makes it testable.

It knows nothing about who lives in the house: there is no register to read, so every line it writes is written
for anybody (no-setup spec)."""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone
from typing import Any, Iterable, Optional
from zoneinfo import ZoneInfo

from sos import conditions as cond
from sos import sun
from sos.rules import Rule, Rules, parse_duration
from sos.system import THEMES

BUCKETS = ("now", "hour", "today", "week")
PHASE_TITLES = {"right-now": "Right now", "first-72-hours": "First 72 hours", "first-month": "First month",
                "long-term": "Long term"}
LINK_KINDS = {"playbook": "playbook", "module": "module", "page": "page", "card": "card", "doc": "doc", "map": "map",
              "kiwix": "kiwix"}
DEFAULT_MODES = {"theme": None, "dim": False, "calls": "shown", "map_first": False, "board": False}
DEFAULT_HOME: dict[str, Any] = {"lat": None, "lon": None, "label": "Home", "flood_zone": None}
FALLBACK_LATLON = (54.0, -2.0)               # the middle of the country until the house is set on the map
WINTER_MONTHS = {11, 12, 1, 2, 3}
SUMMER_MONTHS = {5, 6, 7, 8, 9}


@dataclass
class Model:
    """Everything the engine is allowed to know. Built from the database by the router, or by hand in a test."""
    now: datetime
    conditions: dict[str, cond.Condition] = field(default_factory=dict)
    scenario: Optional[dict] = None                       # {slug, title, started_at, elapsed_s, phase}
    home: dict = field(default_factory=lambda: dict(DEFAULT_HOME))
    drill: bool = False
    checklist: tuple[dict, ...] = ()                      # {id, text} for the active scenario
    checklist_state: dict[str, bool] = field(default_factory=dict)
    kits: tuple[dict, ...] = ()                           # {slug, title, relevant, basic_done, basic_total}
    task_state: dict[str, dict] = field(default_factory=dict)
    detected: dict[str, dict] = field(default_factory=dict)   # condition id -> {state, at, confidence, sensor}
    titles: dict[str, str] = field(default_factory=dict)      # content link -> title, for the briefing
    tz: str = "Europe/London"

    def __post_init__(self) -> None:
        if self.now.tzinfo is None:
            self.now = self.now.replace(tzinfo=timezone.utc)
        self.conditions = {cid: self.conditions.get(cid) or cond.Condition(cid) for cid in cond.IDS}
        self.home = {**DEFAULT_HOME, **(self.home or {})}

    @property
    def latlon(self) -> tuple[float, float]:
        lat, lon = self.home.get("lat"), self.home.get("lon")
        return (float(lat), float(lon)) if lat is not None and lon is not None else FALLBACK_LATLON


# --- matching -----------------------------------------------------------------------------------------------

def phones_state(states: dict[str, str]) -> str:
    """`working` while any phone route is up (a degraded mobile still reaches 999), `off` when neither is."""
    return "off" if states.get("mobile") == "off" and states.get("landline") == "off" else "working"


def _season(now: datetime) -> str:
    return "winter" if now.month in WINTER_MONTHS else ("summer" if now.month in SUMMER_MONTHS else "")


def matches(clause: dict, states: dict[str, str], model: Model, dark: bool) -> bool:
    """Every key in a `when` (or `until`) must hold. An empty clause always holds."""
    for key, value in (clause or {}).items():
        if key in cond.IDS:
            wanted = value if isinstance(value, list) else [value]
            if states.get(key) not in wanted:
                return False
        elif key == "phones":
            if phones_state(states) != value:
                return False
        elif key == "scenario":
            slug = (model.scenario or {}).get("slug")
            if value == "any":
                if slug is None:
                    return False
            elif slug != value:
                return False
        elif key == "dark":
            if bool(value) != dark:
                return False
        elif key == "drill":
            if bool(value) != bool(model.drill):
                return False
        elif key == "season":
            if _season(model.now) != value:
                return False
        else:
            return False
    return True


def trigger_since(clause: dict, model: Model, states: dict[str, str]) -> datetime:
    """When the situation the rule is about began: the most recent `since` among the conditions it names."""
    candidates: list[datetime] = []
    for key, value in (clause or {}).items():
        ids = [key] if key in cond.IDS else (["mobile", "landline"] if key == "phones" else [])
        for cid in ids:
            since = cond.parse_iso(model.conditions[cid].since)
            if since is not None and states.get(cid) != "working":
                candidates.append(since)
        if key == "scenario" and model.scenario:
            started = cond.parse_iso(model.scenario.get("started_at"))
            if started is not None:
                candidates.append(started)
    return max(candidates) if candidates else model.now


# --- the parts of the View ----------------------------------------------------------------------------------

def effective_states(model: Model) -> dict[str, str]:
    """Manual beats detected beats inferred. A detected reading never overrides a hand-set state."""
    states = {}
    for cid, condition in model.conditions.items():
        reading = model.detected.get(cid)
        if reading and condition.source != "manual":
            states[cid] = reading["state"]
        else:
            states[cid] = condition.state
    return states


def condition_view(model: Model, states: dict[str, str]) -> dict[str, dict]:
    out = {}
    for cid, condition in model.conditions.items():
        item = condition.as_dict()
        item["state"] = states[cid]
        item["stale"] = cond.is_stale(condition, model.now)
        if states[cid] == condition.state:
            item["for_s"] = cond.duration_s(condition, model.now)
        else:                                              # a sensor reading is standing in for the stored row
            at = cond.parse_iso((model.detected.get(cid) or {}).get("at")) or model.now
            item["for_s"] = 0 if states[cid] == "working" else max(0, int((model.now - at).total_seconds()))
            item["since"] = at.isoformat() if states[cid] != "working" else None
            item["source"] = "detected"
        out[cid] = item
    return out


def inferred(model: Model, rules: Rules, states: dict[str, str], dark: bool) -> list[dict]:
    """Proposals: what the implications say has probably happened, and what the sensors have noticed."""
    out: list[dict] = []
    for rule in rules.implications:
        if not matches(rule.when, states, model, dark):
            continue
        target = rule.expect["condition"]
        wanted = rule.expect["state"]
        current = states.get(target)
        if current == wanted or (wanted == "degraded" and current == "off") or current is None:
            continue
        due_at = trigger_since(rule.when, model, states) + parse_duration(rule.expect["after"])
        if due_at > model.now:
            continue
        why = rule.why + (f" Applies to {rule.where}." if rule.where else "")
        out.append({"condition": target, "state": wanted, "confidence": rule.confidence,
                    "due_at": due_at.isoformat(), "why": why, "rule": rule.id, "source": rule.source})
    out = _strongest_per_condition(out)
    for cid, reading in sorted(model.detected.items()):
        condition = model.conditions[cid]
        newer = cond.parse_iso(reading.get("at")) or model.now
        updated = cond.parse_iso(condition.updated_at)
        if condition.source == "manual" and reading["state"] != condition.state and (updated is None or newer > updated):
            out.append({"condition": cid, "state": reading["state"], "confidence": float(reading.get("confidence", 0.6)),
                        "due_at": newer.isoformat(), "why": f"The box has detected this itself ({reading.get('sensor', 'sensor')}).",
                        "rule": f"sensor:{reading.get('sensor', cid)}", "source": "page:what-still-works"})
    return out


def _strongest_per_condition(proposals: list[dict]) -> list[dict]:
    """One proposal per condition: `off` beats `degraded`, then the one that came due first."""
    best: dict[str, dict] = {}
    for item in proposals:
        rank = (cond.STATES.index(item["state"]), item["due_at"])
        current = best.get(item["condition"])
        if current is None or rank > (cond.STATES.index(current["state"]), current["due_at"]):
            best[item["condition"]] = item
    return [item for item in proposals if best.get(item["condition"]) is item]


def forecast(model: Model, rules: Rules, states: dict[str, str], dark: bool) -> list[dict]:
    items: list[dict] = []
    for rule in rules.consequences:
        if not matches(rule.when, states, model, dark):
            continue
        since = trigger_since(rule.when, model, states)
        due_at = since + rule.after_td
        items.append({"id": rule.id, "title": rule.title, "due_at": due_at.isoformat(), "severity": rule.severity,
                      "why": rule.why, "link": rule.link, "passed": due_at <= model.now, "rule": rule.id})
    items.sort(key=lambda i: (i["due_at"], i["title"]))
    return items


def _task(model: Model, task_id: str, title: str, bucket: str, why: str, link: Optional[str], source: str,
          done_from_checklist: Optional[bool] = None) -> dict:
    state = model.task_state.get(task_id) or {}
    done = bool(state.get("done")) if done_from_checklist is None else bool(done_from_checklist)
    return {"id": task_id, "title": title, "bucket": bucket, "why": why, "link": link,
            "person": state.get("person") or None, "done": done, "done_at": state.get("done_at"), "source": source}


def task_rule_applies(rule: Rule, model: Model, states: dict[str, str], dark: bool) -> bool:
    if not matches(rule.when, states, model, dark):
        return False
    if rule.until and matches(rule.until, states, model, dark):
        return False
    return not (rule.after and trigger_since(rule.when, model, states) + rule.after_td > model.now)


def tasks(model: Model, rules: Rules, states: dict[str, str], dark: bool) -> list[dict]:
    out: list[dict] = []
    seen: set[str] = set()

    def add(task: dict) -> None:
        if task["id"] not in seen:
            seen.add(task["id"])
            out.append(task)

    for rule in rules.tasks:
        if not task_rule_applies(rule, model, states, dark):
            continue
        add(_task(model, rule.id, rule.title, rule.bucket, rule.why, rule.link, f"rule:{rule.id}"))
    if model.scenario:
        slug = model.scenario["slug"]
        title = model.scenario.get("title") or slug
        for item in model.checklist:
            add(_task(model, f"checklist:{slug}/{item['id']}", item["text"], "today",
                      f"On the {title} checklist.", f"playbook:{slug}#checklist", f"checklist:{slug}",
                      done_from_checklist=bool(model.checklist_state.get(item["id"]))))
    out.sort(key=lambda t: (BUCKETS.index(t["bucket"]) if t["bucket"] in BUCKETS else len(BUCKETS), t["title"]))
    return out


def _briefing_entry(model: Model, link: str) -> dict:
    scheme, _, rest = link.partition(":")
    slug, hash_, fragment = rest.partition("#")
    kind = LINK_KINDS.get(scheme, scheme)
    if scheme == "playbook" and fragment:
        kind = "playbook-section"
    title = model.titles.get(link) or model.titles.get(f"{scheme}:{slug}")
    if not title and scheme == "map":
        title = "The map"
    if not title:
        title = PHASE_TITLES.get(fragment) if fragment else None
    if not title:
        title = slug.replace("-", " ").capitalize() if slug else "The map"
    return {"title": title, "kind": kind, "ref": rest if scheme != "map" else rest.lstrip("?")}


def briefing(model: Model, rules: Rules, states: dict[str, str], dark: bool) -> list[dict]:
    out: list[dict] = []
    seen: set[tuple[str, str]] = set()

    def add(entry: dict) -> None:
        key = (entry["kind"], entry["ref"])
        if key not in seen:
            seen.add(key)
            out.append(entry)

    if model.scenario:
        slug, phase = model.scenario["slug"], model.scenario.get("phase") or "right-now"
        add({"title": PHASE_TITLES.get(phase, phase), "kind": "playbook-section", "ref": f"{slug}#{phase}"})
    for rule in rules.readings:
        if matches(rule.when, states, model, dark):
            for link in rule.open:
                add(_briefing_entry(model, link))
    return out


def modes(model: Model, rules: Rules, states: dict[str, str], dark: bool) -> dict:
    out = dict(DEFAULT_MODES)
    for rule in rules.modes:
        if matches(rule.when, states, model, dark):
            out.update(rule.set or {})
    # A rules file written for the three-theme box still says `theme: vault`, and rules are content:
    # they travel on a stick and outlive the build that reads them. A theme the box has no palette
    # for, stamped on <html>, is a document with no colours at all — so it is no theme, and the
    # reader keeps whichever of the two they were on.
    if out.get("theme") not in THEMES:
        out["theme"] = None
    return out


def next_bulletin(model: Model, rules: Rules) -> dict:
    """The next scheduled bulletin today or tomorrow, in the box's local zone."""
    try:
        zone = ZoneInfo(model.tz)
    except Exception:                                     # an unknown zone on a fresh box: fall back to UTC
        zone = timezone.utc
    local_now = model.now.astimezone(zone)
    best: Optional[tuple[datetime, dict]] = None
    for bulletin in rules.bulletins:
        for text in bulletin.get("times") or []:
            hour, minute = (int(x) for x in text.split(":"))
            for day in (0, 1):
                naive = (local_now + timedelta(days=day)).replace(hour=hour, minute=minute, second=0, microsecond=0)
                at = naive.astimezone(timezone.utc)
                if at <= model.now:
                    continue
                if best is None or at < best[0] or (at == best[0] and bulletin["station"] < best[1]["station"]):
                    best = (at, bulletin)
    if best is None:
        return {"next": None}
    at, bulletin = best
    return {"next": {"station": bulletin["station"], "frequency": bulletin.get("frequency", ""),
                     "at": at.isoformat(), "note": bulletin.get("note", "")}}


def compute(model: Model, rules: Rules) -> dict:
    """The View: everything a screen needs, from one model and one set of rules."""
    lat, lon = model.latlon
    sunrise, sunset = sun.sun_times(lat, lon, model.now)
    sunrise = sunrise.replace(microsecond=0) if sunrise else None
    sunset = sunset.replace(microsecond=0) if sunset else None
    dark = sun.is_dark(lat, lon, model.now)
    states = effective_states(model)
    return {
        "meta": {"now": model.now.isoformat(), "dark": dark,
                 "sunrise": sunrise.isoformat() if sunrise else None,
                 "sunset": sunset.isoformat() if sunset else None,
                 "home": dict(model.home), "drill": bool(model.drill)},
        "scenario": dict(model.scenario) if model.scenario else None,
        "conditions": condition_view(model, states),
        "inferred": inferred(model, rules, states, dark),
        "forecast": forecast(model, rules, states, dark),
        "tasks": tasks(model, rules, states, dark),
        "briefing": briefing(model, rules, states, dark),
        "modes": modes(model, rules, states, dark),
        "bulletins": next_bulletin(model, rules),
    }


def flags_for(states: dict[str, str], scenario: Optional[str], dark: bool) -> dict[str, bool]:
    """The content directives' flag set: a service counts as working while it is not off."""
    out = {cid: states.get(cid, "working") != "off" for cid in cond.IDS}
    out["phones"] = out["mobile"] or out["landline"]
    out["dark"] = bool(dark)
    if scenario:
        out[f"scenario:{scenario}"] = True
    return out


def flags(view: dict) -> dict[str, bool]:
    """The flag set for a computed View."""
    return flags_for({cid: item["state"] for cid, item in view["conditions"].items()},
                     (view.get("scenario") or {}).get("slug"), view["meta"]["dark"])


def summary_line(view: dict) -> str:
    """One line for the log and the report: what is not working."""
    broken = [item["title"] for item in view["conditions"].values() if item["state"] != "working"]
    return ", ".join(broken) if broken else "Everything working"


def report(view: dict, events: Iterable[dict] = ()) -> str:
    """The printable hand-over: the situation, the forecast, the tasks and the log, as Markdown."""
    lines: list[str] = ["# Situation report", ""]
    meta, scenario = view["meta"], view.get("scenario")
    lines.append(f"Written {meta['now']}." + (" **This is a drill.**" if meta.get("drill") else ""))
    lines.append("")
    if scenario:
        hours = scenario.get("elapsed_s", 0) / 3600
        lines += [f"**{scenario.get('title') or scenario['slug']}**, started {scenario['started_at']} "
                  f"({hours:.1f} hours ago, phase: {scenario.get('phase')}).", ""]
    lines += ["## Conditions", ""]
    for item in view["conditions"].values():
        note = f" — {item['note']}" if item.get("note") else ""
        since = f" since {item['since']}" if item.get("since") else ""
        lines.append(f"- **{item['title']}**: {item['state']}{since} ({item['source']}){note}")
    if view["inferred"]:
        lines += ["", "## Probably also", ""]
        for item in view["inferred"]:
            lines.append(f"- {item['condition']} may be {item['state']} ({int(item['confidence'] * 100)}%): {item['why']}")
    if view["forecast"]:
        lines += ["", "## What is coming", ""]
        for item in view["forecast"]:
            lines.append(f"- {'PASSED' if item['passed'] else 'due'} {item['due_at']}: **{item['title']}** ({item['severity']}) — {item['why']}")
    lines += ["", "## Tasks", ""]
    for bucket in BUCKETS:
        rows = [t for t in view["tasks"] if t["bucket"] == bucket]
        if not rows:
            continue
        lines.append(f"### {bucket.capitalize()}")
        for task in rows:
            who = f" — {task['person']}" if task.get("person") else ""
            lines.append(f"- [{'x' if task['done'] else ' '}] {task['title']}{who}")
        lines.append("")
    if view["briefing"]:
        lines += ["## Read", ""] + [f"- {b['title']} ({b['kind']} {b['ref']})" for b in view["briefing"]] + [""]
    events = list(events)
    if events:
        lines += ["## Log", ""] + [f"- {e.get('updated_at')}: {e.get('title')}" for e in events] + [""]
    return "\n".join(lines).rstrip() + "\n"

File: web/e2e/fixtures/engine.ts
// A small in-memory situation engine for the Playwright fixtures: enough of the contract for the
// screens to be exercised without the backend (conditions, a fixed forecast when the power is off,
// tasks with their done state, modes, briefing, readiness).
import { CONDITION_IDS, type Condition, type ConditionId, type ConditionState, type Conditions, type Forecast, type Inferred, type SituationView, type Task } from '../../src/api/types';
import { phaseFor } from '../../src/tools/situation';
import type { FixtureState } from './state';

const HOUR = 3_600_000;
const TITLES: Record<ConditionId, string> = {
  power: 'Mains power', water: 'Water supply', mobile: 'Mobile network', landline: 'Landline and 999', internet: 'Internet',
  gas: 'Gas', heating: 'Heating', roads: 'Roads and transport', shops: 'Shops and cash', sewage: 'Sewage and drains',
};

export function freshConditions(at = new Date().toISOString()): Conditions {
  return Object.fromEntries(CONDITION_IDS.map((id) => [id, {
    id, title: TITLES[id], state: 'working' as ConditionState, since: at, for_s: 0, source: 'manual' as const,
    confidence: 1, note: '', set_by: 'phone', updated_at: at, confirmed_at: at, stale: false,
  }])) as Conditions;
}

function aged(c: Condition, now: number): Condition {
  const since = c.since ? Date.parse(c.since) : now;
  const for_s = Math.max(0, Math.round((now - since) / 1000));
  const confirmed = c.confirmed_at ? Date.parse(c.confirmed_at) : Date.parse(c.updated_at);
  return { ...c, for_s, stale: c.state !== 'working' && now - confirmed >= 24 * HOUR };
}

const sinceOf = (c: Condition): number => (c.since ? Date.parse(c.since) : Date.now());

function task(id: string, title: string, bucket: Task['bucket'], why: string, link: string, state: FixtureState): Task {
  const saved = state.taskState.get(id);
  return { id, title, bucket, why, link, person: saved?.person ?? null, done: saved?.done ?? false, done_at: saved?.done_at ?? null, source: `rule:${id}` };
}

/** The gaps come from the same register and the same stock the household summary counts, so the
 * front door cannot say two different things about the same water. */
function readinessGaps(state: FixtureState): { title: string; link: string; points: number }[] {
  const gaps: { title: string; link: string; points: number }[] = [];
  const people = Math.max(1, state.household.length);
  const water = state.stock.filter((s) => s.category === 'water');
  const food = state.stock.filter((s) => s.category === 'food');
  const daysOf = (items: typeof state.stock) => items.reduce((n, i) => n + (i.days_left ?? 0), 0);
  if (state.household.length === 0) gaps.push({ title: 'Say who lives here', link: '/plan#household', points: 12 });
  if (water.length === 0) gaps.push({ title: 'No water recorded: add what you have', link: '/plan#stock', points: 12 });
  else if (daysOf(water) < 3) gaps.push({ title: `Water: ${daysOf(water).toFixed(1)} days for ${people} ${people === 1 ? 'person' : 'people'}`, link: '/plan#stock', points: 12 });
  if (food.length === 0) gaps.push({ title: 'No food recorded: add what you have', link: '/plan#stock', points: 8 });
  return gaps;
}

export function computeView(state: FixtureState, now = Date.now()): SituationView {
  const conditions = Object.fromEntries(CONDITION_IDS.map((id) => [id, aged(state.conditions[id], now)])) as Conditions;
  const off = (id: ConditionId) => conditions[id].state === 'off';
  const broken = (id: ConditionId) => conditions[id].state !== 'working';
  const power = conditions.power;

  const inferred: Inferred[] = [];
  if (off('power') && power.for_s >= 4 * 3600 && conditions.mobile.state === 'working') {
    inferred.push({
      condition: 'mobile', state: 'degraded', confidence: 0.7, due_at: new Date(sinceOf(power) + 4 * HOUR).toISOString(),
      why: 'Masts run about four hours on battery once the street power goes.', rule: 'power-off-mobile-degraded', source: 'page:what-still-works',
    });
  }

  // A reading never overrides what somebody said; it proposes, flagged as the box's own detection.
  const internet = state.sensors.internet;
  if (internet && internet.value === 0 && conditions.internet.state === 'working') {
    inferred.push({
      condition: 'internet', state: 'off', confidence: 0.8, due_at: internet.at, detected: true,
      why: 'The box has not reached the internet since ' + internet.at.slice(11, 16) + '.',
      rule: 'sensor:internet', source: 'page:what-still-works',
    });
  }

  const forecast: Forecast[] = [];
  if (off('power')) {
    const from = sinceOf(power);
    const item = (id: string, title: string, hours: number, severity: Forecast['severity'], why: string): Forecast => {
      const due = new Date(from + hours * HOUR).toISOString();
      return { id, title, due_at: due, severity: Date.parse(due) <= now ? 'passed' : severity, why, link: 'module:food', passed: Date.parse(due) <= now };
    };
    forecast.push(item('fridge', 'Fridge food unsafe', 4, 'warn', 'A closed fridge holds about four hours.'));
    forecast.push(item('phones', 'Phone batteries flat', 12, 'warn', 'Charge everything while the power lasts.'));
    forecast.push(item('freezer', 'Freezer food unsafe', 24, 'danger', 'A half-full freezer holds about 24 hours; a full one, 48.'));
    forecast.sort((a, b) => a.due_at.localeCompare(b.due_at));
  }

  const tasks: Task[] = [];
  if (off('power')) {
    if (!off('water')) tasks.push(task('fill-bath', 'Fill the bath and every container', 'now', 'Pumped supplies fail about a day after the power does.', 'module:water', state));
    tasks.push(task('freezer-shut', 'Keep the fridge and freezer doors shut', 'now', 'Every opening costs hours of cold.', 'module:food', state));
    tasks.push(task('cooker-off', 'Turn the cooker off at the knobs', 'now', 'It will come back on with the power.', 'module:power', state));
    tasks.push(task('cash', 'Get cash out while the shops can take it', 'hour', 'Card terminals need power.', 'page:what-still-works', state));
  }
  if (broken('water')) tasks.push(task('boil-water', 'Boil water before drinking it', 'today', 'Pressure loss lets dirt into the pipes.', 'page:water-disinfection', state));
  if (off('mobile') && off('landline')) tasks.push(task('meeting-point', 'Agree a meeting point and a runner', 'now', 'Nobody can ring anybody.', 'page:no-phones', state));
  if (state.situation.slug) {
    const list = state.checklists.get(state.situation.slug) ?? [];
    for (const item of list) {
      const id = `checklist:${state.situation.slug}/${item.id}`;
      tasks.push({ id, title: item.text, bucket: 'today', why: 'From this situation’s checklist.', link: `playbook:${state.situation.slug}`, person: state.taskState.get(id)?.person ?? null, done: item.checked, done_at: item.updated_at, source: `checklist:${state.situation.slug}` });
    }
  }
  // Phase 4: a neighbour who needs checking on is a job like any other, with the same id on both sides.
  const slug = (name: string) => name.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/^-|-$/g, '');
  const checkOn = (off('power') || off('water') || off('mobile'))
    ? state.neighbours.filter((n) => n.needs.trim()).map((n) => ({
        id: `neighbour:${slug(n.name)}`, name: n.name, address: n.address, needs: n.needs, contacts: n.contacts,
        title: `Knock on ${n.name}${n.address ? `, ${n.address}` : ''}`, why: n.needs,
        rule: `neighbour-check-${slug(n.name)}`, link: null, bucket: 'today' as const,
        done: state.taskState.get(`neighbour:${slug(n.name)}`)?.done ?? false,
      }))
    : [];
  const skills = state.neighbours.filter((n) => n.skills.trim()).map((n) => ({
    name: n.name, address: n.address, skill: n.skills, text: n.skills, contacts: n.contacts,
    why: 'On the street list.', rule: `neighbour-skill-${slug(n.name)}`, link: null,
  }));
  for (const c of checkOn) {
    tasks.push({ id: c.id, title: c.title, bucket: c.bucket, why: c.why, link: c.link, person: state.taskState.get(c.id)?.person ?? null, done: c.done, done_at: null, source: 'rule:neighbours' });
  }

  const order = { now: 0, hour: 1, today: 2, week: 3 };
  tasks.sort((a, b) => order[a.bucket] - order[b.bucket] || a.title.localeCompare(b.title));

  const briefing: SituationView['briefing'] = [];
  if (state.situation.slug) briefing.push({ title: 'Right now', kind: 'playbook-section', ref: `${state.situation.slug}#right-now` });
  if (broken('power')) {
    briefing.push({ title: 'Power', kind: 'module', ref: 'power' });
    briefing.push({ title: 'What still works in an outage', kind: 'page', ref: 'what-still-works' });
  }
  if (broken('water')) briefing.push({ title: 'Water', kind: 'module', ref: 'water' });

  const callsHidden = off('mobile') && off('landline');
  const scenario = state.situation.slug
    ? { slug: state.situation.slug, title: state.situation.title ?? state.situation.slug, started_at: state.situation.started_at, elapsed_s: Math.max(0, Math.round((now - Date.parse(state.situation.started_at)) / 1000)), phase: phaseFor(Math.max(0, (now - Date.parse(state.situation.started_at)) / 1000)).id }
    : null;

  return {
    meta: {
      now: new Date(now).toISOString(), dark: state.dark,
      sunrise: '2026-09-06T05:22:00.000Z', sunset: '2026-09-06T18:41:00.000Z',
      home: state.home, drill: state.drill,
    },
    scenario,
    conditions,
    inferred,
    forecast,
    tasks,
    briefing,
    modes: {
      theme: off('power') && state.dark ? 'mono' : null,
      dim: off('power') && state.dark,
      calls: callsHidden ? 'hidden' : 'shown',
      map_first: state.situation.slug === 'storms-flooding',
      board: Boolean(state.situation.slug),
    },
    readiness: { score: 62, gaps: readinessGaps(state) },
    bulletins: { next: { station: 'BBC Radio 4', frequency: '198 kHz LW', at: '2026-09-06T18:00:00.000Z' } },
    neighbours: { check_on: checkOn, skills },
  };
}

/** The hand-over sheet: what the print button opens. */
export function report(state: FixtureState): string {
  const view = computeView(state);
  const lines = ['# Situation report', '', `As at ${view.meta.now}.`, '', '## Conditions', ''];
  for (const id of CONDITION_IDS) lines.push(`- ${view.conditions[id].title}: ${view.conditions[id].state}`);
  lines.push('', '## Tasks', '');
  for (const t of view.tasks) lines.push(`- [${t.done ? 'x' : ' '}] ${t.title}`);
  return `${lines.join('\n')}\n`;
}
